app/master/job_config.py

```diff
@@ -18,3 +35,119 @@ def __init__(self, name, setup_build, teardown_build, command, atomizer, max_exe
+        :type config_dict: dict
+        :rtype: None
+        """
+        required_fields = {COMMANDS, ATOMIZERS}
+        allowed_fields_expected_types = {
+            SETUP_BUILD: [(list, str)],  # (list, str) means this field should be a list of strings
+            TEARDOWN_BUILD: [(list, str)],
+            COMMANDS: [(list, str)],
+            ATOMIZERS: [(list, dict)],  # (list, dict) means this field should be a list of dicts
+            MAX_EXECUTORS: [int],
+            MAX_EXECUTORS_PER_SLAVE: [int],
+        }
+
+        if not isinstance(config_dict, dict):
+            raise ConfigValidationError('Passed in configuration is not a dictionary for job: "{}".'.format(name))
+
+        missing_required_fields = required_fields - config_dict.keys()
+        if missing_required_fields:
+            raise ConfigValidationError('Definition for job "{}" is missing required config sections: {}'
+                                        .format(name, missing_required_fields))
+
+        for config_section_name, config_section_value in config_dict.items():
+            if config_section_name not in allowed_fields_expected_types:
+                raise ConfigValidationError('Definition for job "{}" contains an invalid config section "{}".'
+                                            .format(name, config_section_name))
+
+            expected_section_types = allowed_fields_expected_types[config_section_name]
+            actual_section_type = type(config_section_value)
+            if actual_section_type is list:
+                # also check the type of the list items (assuming all list items have the same type as the first)
+                actual_section_type = (list, type(config_section_value[0]))
+
+            if actual_section_type not in expected_section_types:
+                raise ConfigValidationError(
+                    'Definition for job "{}" contains an invalid value for config section "{}". '
+                    'Parser expected one of {} but found {}.'
+                    .format(name, config_section_name, expected_section_types, actual_section_type))
+
+    @classmethod
+    def _unpack(cls, name, config_dict):
+        """
+        Set class attributes from config dictionary.
+
+        :type name: str
+        :type config_dict: dict
+        :rtype: JobConfig
+        """
+        setup_build = cls._shell_command_list_to_single_command(config_dict.get(SETUP_BUILD))
+        teardown_build = cls._shell_command_list_to_single_command(config_dict.get(TEARDOWN_BUILD))
+        command = cls._shell_command_list_to_single_command(config_dict[COMMANDS])
+        atomizer = Atomizer(config_dict[ATOMIZERS])
+        max_executors = config_dict.get(MAX_EXECUTORS, cls.DEFAULT_MAX_EXECUTORS)
+        max_executors_per_slave = config_dict.get(MAX_EXECUTORS_PER_SLAVE, cls.DEFAULT_MAX_EXECUTORS)
+        return JobConfig(name, setup_build, teardown_build, command, atomizer, max_executors, max_executors_per_slave)
```

Comment:
Minor thing, but I think you can do `return cls(name, setup_build, ...)` here instead of reference the containing classname explicitly.